Runtime/Scripts/Timeline/GLTFRecorder.cs

```diff
@@ -269,148 +343,143 @@ private static void SafeDestroy(Object obj)
+					collectAndProcessSingleTrack(
+						gltfSceneExporter,
+						anim,
+						track,
+						visibilityTrack,
+						calculateTranslationBounds,
+						ref gotFirstValue,
+						ref translationBounds,
+						ref weHadAScaleTrack
+					);
+				}
 
-					var animatedObject = track.AnimatedObject;
-					var trackName = track.PropertyName;
-					var trackTimes = track.Times;
-					var trackValues = track.Values;
-					
-					// AnimationData always has a visibility track, and if there is also a scale track, merge them
-					if (track.PropertyName == "scale" && track is AnimationTrack<Transform, Vector3> scaleTrack) {
-						// GLTF does not internally support a visibility state (animation).
-						// So to simulate support for that, merge the visibility track with the scale track
-						// forcing the scale to (0,0,0) whenever the model is invisible
-						var (mergedTimes, mergedScales) = mergeVisibilityAndScaleTracks(kvp.Value.visibilityTrack, scaleTrack);
-						if (mergedTimes == null || mergedScales == null)
+				if (visibilityTrack != null && !weHadAScaleTrack) {
+					if (visibilityTrack.Values.Length <= 2) {
+						// use nullable to be able to distinguish "no value" from "first value is invisible"
+						bool? first = visibilityTrack.Values.Length > 0 ? visibilityTrack.Values[0] : null;
+						if (first == null || visibilityTrack.Values.All(v => first == v)) 
 							continue;
-						trackTimes = mergedTimes;
-						trackValues = mergedScales.Cast<object>().ToArray();
 					}
-					OnBeforeAddAnimationData?.Invoke(new PostAnimationData(animatedObject, trackName, trackTimes, trackValues));
-
-					if (calculateTranslationBounds && track.PropertyName == "translation") {
-						
-						foreach (var t in trackValues) {
-							var vec = (Vector3) t;
-							if (!gotFirstValue)
-							{
-								translationBounds = new Bounds(vec, Vector3.zero);
-								gotFirstValue = true;
-							}
-							else
-							{
-								translationBounds.Encapsulate(vec);
-							}
-						}
-					}
-
-					GLTFSceneExporter.RemoveUnneededKeyframes(ref trackTimes, ref trackValues);
-					gltfSceneExporter.AddAnimationData(track.AnimatedObject, track.PropertyName, anim, trackTimes, trackValues);
+					
+					var (interpolation, times, scales) = visibilityTrackToScaleTrack(visibilityTrack);
+					gltfSceneExporter.AddAnimationData(kvp.Key, "scale", anim, interpolation, times, scales.Cast<object>().ToArray());
 				}
-				processAnimationMarker.End();
 			}
 		}
 
-		private static (double[]? times, Vector3[]? mergedScales) mergeVisibilityAndScaleTracks(
+		private void collectAndProcessSingleTrack(
+			AnimationDataCollector gltfSceneExporter,
+			GLTFAnimation animation,
+			AnimationTrack track,
 			VisibilityTrack? visibilityTrack,
-			BaseAnimationTrack<Transform, Vector3>? scaleTrack
+			bool calculateTranslationBounds,
+			ref bool foundFirstTranslationTrack,
+			ref Bounds translationBounds,
+			ref bool foundScaleTrack
```

Comment:
I'm not a big fan of reafs, are these really needed?

Reply:
in this case they are kind of needed - or i would have to return these three each call